firefox-ios/Client/Frontend/Browser/Toolbars/AddressBarPanGestureHandler.swift

```diff
@@ -0,0 +1,151 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import UIKit
+import Common
+
+final class AddressBarPanGestureHandler: NSObject {
+    // MARK: - UX Constants
+    private struct UX {
+        static let offset: CGFloat = 24
+        static let swipingDuration: TimeInterval = 0.2
+        static let swipingVelocity: CGFloat = 150
+    }
+
+    // MARK: - UI Properties
+    private let contentContainer: ContentContainer
+    private let webPagePreview: TabWebViewPreview
+    private var originalPosition = CGPoint()
+    private var panGestureRecognizer: UIPanGestureRecognizer?
+    private var addressBarContainer: BaseAlphaStackView
+
+    // MARK: - Properties
+    private let tabManager: TabManager
+    private let windowUUID: WindowUUID
+    private let screenshotHelper: ScreenshotHelper?
+
+    // MARK: - Init
+    init(
+        contentContainer: ContentContainer,
+        addressBarContainer: BaseAlphaStackView,
+        webPagePreview: TabWebViewPreview,
+        tabManager: TabManager,
+        windowUUID: WindowUUID,
+        screenshotHelper: ScreenshotHelper?
+    ) {
+        self.contentContainer = contentContainer
+        self.addressBarContainer = addressBarContainer
+        self.webPagePreview = webPagePreview
+        self.tabManager = tabManager
+        self.windowUUID = windowUUID
+        self.screenshotHelper = screenshotHelper
+        super.init()
+        updateAddressBarContainer(addressBarContainer)
+    }
+
+    /// Updates the address bar container with a new container view and reattaches the pan gesture recognizer.
+    ///
+    /// - Parameter newContainer: The new `BaseAlphaStackView` to be used as the address bar container.
+    func updateAddressBarContainer(_ newContainer: BaseAlphaStackView) {
+        if let panGestureRecognizer {
+            addressBarContainer.removeGestureRecognizer(panGestureRecognizer)
+        }
+        addressBarContainer = newContainer
+
+        let newPanGesture = UIPanGestureRecognizer(target: self, action: #selector(handlePan))
+        addressBarContainer.addGestureRecognizer(newPanGesture)
+
+        panGestureRecognizer = newPanGesture
+    }
+
+    // MARK: - Pan Gesture Handling
+    @objc
+    private func handlePan(_ gesture: UIPanGestureRecognizer) {
+        let translation = gesture.translation(in: contentContainer)
+
+        guard let selectedTab = tabManager.selectedTab else { return }
+        let tabs = selectedTab.isPrivate ? tabManager.privateTabs : tabManager.normalTabs
+        guard let index = tabs.firstIndex(where: { $0 === selectedTab }) else { return }
+
+        switch gesture.state {
+        case .began:
+            originalPosition = contentContainer.frame.origin
+            // Set the initial position of webPagePreview with the offset
+            webPagePreview.frame.origin.x = translation.x < 0
+            ? originalPosition.x + UX.offset
+            : -originalPosition.x - UX.offset
```

Comment:
nit: A ternary operator should never go over multiple lines, it defeats the purpose of it and makes it hard to read. The same goes for the code later in the class.
Also I don't know if it was ever documented anywhere but we discussed a while back trying to keep case statements to a max of a couple of lines. A small helper function here to calculate the offset would solve both issues in this instance.

Reply:
Added a helper function, modified some parts of the code for better readability and I promise to write tests on a follow-up PR.